tidy3d/plugins/design/method.py

```diff
@@ -175,28 +110,466 @@ class MethodGrid(MethodIndependent):
     >>> method = tdd.MethodGrid()
     """
 
-    def sample(self, parameters: Tuple[ParameterType, ...], **kwargs) -> Dict[str, Any]:
+    @staticmethod
+    def sample(parameters: Tuple[ParameterType, ...]) -> Dict[str, Any]:
         """Defines how the design parameters are sampled on grid."""
 
         # sample each dimension individually
         vals_each_dim = {}
-        for design_var in parameters:
-            vals = design_var.sample_grid()
-            vals_each_dim[design_var.name] = vals
+        for param in parameters:
+            vals = param.sample_grid()
+            vals_each_dim[param.name] = vals
 
         # meshgrid each dimension's results and combine them all
         vals_grid = np.meshgrid(*vals_each_dim.values())
         vals_grid = (np.ravel(x).tolist() for x in vals_grid)
-        return dict(zip(vals_each_dim.keys(), vals_grid))
+        vals_dict = dict(zip(vals_each_dim.keys(), vals_grid))
+        t_vals_dict = [dict(zip(vals_dict.keys(), values)) for values in zip(*vals_dict.values())]
 
+        return t_vals_dict
 
-class AbstractMethodRandom(MethodIndependent, ABC):
+
+class MethodOptimize(Method, ABC):
+    """A method for handling design searches that optimize the design"""
+
+    # NOTE: We could move this to the Method base class but it's not relevant to MethodGrid
+    rng_seed: pd.PositiveInt = pd.Field(
+        default=None,
+        title="Seed for random number generation",
+        description="Set the seed used by the optimizers to set constant random number generation.",
+    )
+
+    def any_to_int_param(self, parameter):
+        """Convert ParameterAny object to integers and provide a conversion dict to return"""
+
+        return dict(enumerate(parameter.allowed_values))
+
+    def sol_array_to_dict(
+        self, solution: np.array, keys: list, param_converter: dict
+    ) -> list[dict]:
+        """Convert an array of solutions to a list of dicts for function input"""
+        sol_dict_list = [dict(zip(keys, sol)) for sol in solution]
+
+        self._handle_param_convert(param_converter, sol_dict_list)
+
+        return sol_dict_list
+
+    def _handle_param_convert(self, param_converter, sol_dict_list):
+        for param, convert in param_converter.items():
+            for sol in sol_dict_list:
+                if isinstance(sol[param], float):
+                    sol[param] = int(round(sol[param], 0))
+                sol[param] = convert[sol[param]]
+
+
+class MethodBayOpt(MethodOptimize, ABC):
+    """A standard method for performing bayesian optimization search"""
+
+    initial_iter: pd.PositiveInt = pd.Field(
+        ...,
```

Comment:
is there a good default for this?

Reply:
Not especially, it's quite problem specific. I could set it to something vaguely sensible like 10, but I'd rather the user set it every time. If only because it would be confusing to the user if they didn't realise it had a default.